base/namedtuple.jl

```diff
@@ -112,6 +112,9 @@ NamedTuple{names}(itr) where {names} = NamedTuple{names}(Tuple(itr))
 
 NamedTuple(itr) = (; itr...)
 
+# avoids invalidating Union{}(...)
+NamedTuple{names, Union{}}(itr::Tuple) where {names} = throw(MethodError(NamedTuple{names, Union{}}, (itr,)))
```

Comment:
It seems that the ideal fix here would be
```julia
julia> f(args) = Tuple{args...}
f (generic function with 1 method)

julia> @code_typed f([])
CodeInfo(
1 ─ %1 = Core._apply_iterate(Base.iterate, Core.apply_type, (Tuple,), args)::Type
└──      return %1
) => Type
```
and make this `Tuple` instead of `Type`. Do you understand why it isn't?

Reply:
This invalidation comes from `JuliaInterpreter.namedtuple`, I can see whether this can perhaps be fixed there directly. It's probably more of a general question whether the burden here should just fall on the package authors, or if we still want to work around this in Base, if it's easy enough like this.

Reply:
That's not what I'm suggesting; I'm suggesting the fix lies in `Core.Compiler`, causing `Tuple{container...}` to be inferred as returning a `Tuple` even when `container::Vector{Any}`.

Reply:
Hmm, I tried it now with these changes:
```diff
diff --git a/base/compiler/tfuncs.jl b/base/compiler/tfuncs.jl
index e8057fbe74..2942190c58 100644
--- a/base/compiler/tfuncs.jl
+++ b/base/compiler/tfuncs.jl
@@ -1134,7 +1134,9 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
     else
         return Any
     end
+    istuple = isa(headtype, Type) && (headtype == Tuple)
     if !isempty(args) && isvarargtype(args[end])
+        istuple && return Type{Tuple{Vararg{Any, N}}} where N
         return isvarargtype(headtype) ? Core.TypeofVararg : Type
     end
     largs = length(args)
@@ -1177,7 +1179,6 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
         end
         return allconst ? Const(ty) : Type{ty}
     end
-    istuple = isa(headtype, Type) && (headtype == Tuple)
     if !istuple && !isa(headtype, UnionAll) && !isvarargtype(headtype)
         return Union{}
     end
@@ -1237,9 +1238,9 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
                     # If the names are known, keep the upper bound, but otherwise widen to Tuple.
                     # This is a widening heuristic to avoid keeping type information
                     # that's unlikely to be useful.
-                    if !(uw.parameters[1] isa Tuple || (i == 2 && tparams[1] isa Tuple))
-                        ub = Any
-                    end
+                    #if !(uw.parameters[1] isa Tuple || (i == 2 && tparams[1] isa Tuple))
+                    #    ub = Any
+                    #end
                 else
                     ub = Any
                 end
```
and it does seem to do the correct thing:
```julia
julia> f(x, y, z) = NamedTuple{(x...,),Tuple{y...}}(z...)
f (generic function with 1 method)

julia> @code_warntype f(Any[], Any[], Any[])
MethodInstance for f(::Vector{Any}, ::Vector{Any}, ::Vector{Any})
  from f(x, y, z) in Main at REPL[1]:1
Arguments
  #self#::Core.Const(f)
  x::Vector{Any}
  y::Vector{Any}
  z::Vector{Any}
Body::Any
1 ─ %1 = Core._apply_iterate(Base.iterate, Core.tuple, x)::Tuple
│   %2 = Core.tuple(Main.Tuple)::Core.Const((Tuple,))
│   %3 = Core._apply_iterate(Base.iterate, Core.apply_type, %2, y)::Type{Tuple{Vararg{Any, N}}} where N
│   %4 = Core.apply_type(Main.NamedTuple, %1, %3)::Type{NamedTuple{_A, _B}} where {_A, _B<:(Tuple{Vararg{Any, N}} where N)}
│   %5 = Core._apply_iterate(Base.iterate, %4, z)::Any
└──      return %5
```
I still see the same invalidations though, so fixing this in the `apply_type` tfunc does not seem to be enough.

Reply:
Hmm, too bad. I guess I was bothered about
```julia
julia> NamedTuple{(), Union{}}
ERROR: NamedTuple field type must be a tuple type
```
but
```julia
julia> NamedTuple{names, Union{}} where names
NamedTuple{names, Union{}} where names
```
So it almost seems that nixing this type is the better solution.

But since having it be a `Tuple` doesn't fix the invalidation, and since we allow that `UnionAll` to be constructed, your original version seems fine. Is it fragile to adding the change to `apply_type_tfunc`? Meaning, will having it infer as `<:Tuple{Vararg{Any}}` nix your fix?

Reply:
> Meaning, will having it infer as <:Tuple{Vararg{Any}} nix your fix?

No, this would still apply, since `Union{}` is still a subtype. This also explains why changing `apply_type_tfunc` did not fix anything.

> And do you have to restrict your fix to iter::Tuple?

This would technically cause ambiguities and I am not sure how intentional ambiguities affect invalidations.